From 5bfa85b6633cb8d244cc0bcc09f190f553956197 Mon Sep 17 00:00:00 2001
From: andrei varvara <andrei.varvara@freescale.com>
Date: Fri, 18 May 2012 20:49:43 +0000
Subject: [PATCH 165/518] Made SMP safe the create SA function from DPA IPSec
 code

synchronized the dpa_ipsec_create_sa function

Signed-off-by: Andrei Varvara <andrei.varvara@freescale.com>
Acked-by: Mihai Serb <mihai.serb@freescale.com>
[Grabbed from the branch, LINUX_IR5.2.0, of
https://git.freescale.com/git-private/cgit.cgi/ppc/alu-b4860/linux.git.]
Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
---
 drivers/staging/fsl_dpa_offload/dpa_ipsec.c |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/drivers/staging/fsl_dpa_offload/dpa_ipsec.c b/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
index c6d861f..baba5ba 100644
--- a/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
+++ b/drivers/staging/fsl_dpa_offload/dpa_ipsec.c
@@ -2822,11 +2822,14 @@ int dpa_ipsec_create_sa(int dpa_ipsec_id,
 		return err;
 	}
 
-	/* Update internal SA structure */
+	/* Update internal SA structure. First acquire its lock */
+	mutex_lock(&sa->lock);
 	sa->sa_dir = sa_params->sa_dir;
 	sa->dpa_ipsec = dpa_ipsec;
 	sa->parent_sa = NULL;
 	sa->child_sa = NULL;
+	sa->sa_rekeying_node.next = LIST_POISON1;
+	sa->sa_rekeying_node.prev = LIST_POISON2;
 
 	/* Copy SA params into the internal SA structure */
 	if (sa->sa_dir == DPA_IPSEC_OUTBOUND)
@@ -2920,6 +2923,9 @@ int dpa_ipsec_create_sa(int dpa_ipsec_id,
 	/* SA done ok. Return the SA id */
 	*sa_id = id;
 
+	/* Unlock the SA structure */
+	mutex_unlock(&sa->lock);
+
 	return 0;
 
 	/* Something went wrong. Begin rollback */
@@ -2933,6 +2939,9 @@ int dpa_ipsec_create_sa(int dpa_ipsec_id,
 	if (err_rb < 0)
 		*sa_id = id;
 
+	/* Unlock the SA structure */
+	mutex_unlock(&sa->lock);
+
 	return err;
 }
 EXPORT_SYMBOL(dpa_ipsec_create_sa);
-- 
1.7.5.4

